README.md
# BiAlign - Bialignment of RNAs and proteins

This type of alignment supports evolutionary 'shift' events that allow some incongruence between sequence and structure evolution.

![](Examples/example.svg)

The first version of this tool has been described in 

Waldl M., Will S., Wolfinger M.T., Hofacker I.L., Stadler P.F. (2020)
Bi-alignments as Models of Incongruent Evolution of RNA Sequence and
Secondary Structure. In: Cazzaniga P., Besozzi D., Merelli I., Manzoni L.
(eds) Computational Intelligence Methods for Bioinformatics and
Biostatistics. CIBB 2019. Lecture Notes in Computer Science, vol 12313.
Springer, Cham. https://doi.org/10.1007/978-3-030-63061-4_15


## Setup and usage

The software requires compilation by Cython. For this purpose Cython and
Python need to be installed and the tool has to be compiled by

```
python setup.py build_ext --inplace
```

Moreover, for aligning RNAs, the tool requires the Vienna RNA package with Python bindings. We recommend to use the tool under Linux or MacOS and install the prerequisites via conda / bioconda.

## Usage examples

The tool can be used from the command line or via its Python interface
(e.g. from a Jupyter notebook).
### Command line interface

To get an overview on all command line parameters that configure modes and
alignment parameters, please refer to the help output of the tool as obtained by

```
./bialign.py --help
```

#### RNA bi-alignment examples

This 'toy' example demonstrates a simple helix shift;

```bash
./bialign.py GCGGGGGAUAUCCCCAUCG GGGGAUAUCCCCAUCG \
    --strA "...(((.....)))....." --strB ".(((.....)))...." \
    --structure 400 \
    --gap_opening_cost -200 --gap_cost -50 \
    --max_shift 0 --shift_cost -150
```

Structures will be predicted (using the Vienna RNA package) if they are not
explicitly given, e.g.
```
./bialign.py UGUAAACAUCCUCGACUGGAAGCUGUGAAGCCACAAAUGGGCUUUCAGUCGGAUGUUUGCA UGUAAACAUCCUACACUCAGCUGUCAUACAUGCGUUGGCUGGGAUGUGGAUGUUUACG 
```
Note that this fails, if the Vienna RNA package with Python binding is not
available.


#### Bi-Alignments of proteins with affine gap cost

```bash
./bialign.py RAKLPLKEKKLTATANYHPGIRYIMTGYSAKYIYSSTYARFR KAKLPLKEKKLTRTANYHPGIRYIMTGYSAKRIYSSTYAYFR \
    --strA "CHHHHHHHHHHHHHCCCCTCEEEEEEECCTCEEEEEEEECCC" --strB "HHHHHHHHHHHHCCCCCCTCEEEEEEECCCCCEEEEEEEECC" \
    --type Protein --shift_cost -150 --structure_weight 800 --simmatrix BLOSUM62 --gap_opening_cost -150 \
    --gap_cost -50 --max_shift 1
```

Input can also be read from files as written by the secondary structure prediction
web server CFSSP (Kumar et al, 2013; http://www.biogem.org/tool/chou-fasman).

```bash
./bialign.py --filein Examples/DNAPolymerase1_Escherichia.cfssp Examples/DNAPolymerase1_Xanthomonas.cfssp \
    --type Protein --shift_cost -150 --structure_weight 800 --simmatrix BLOSUM62 --gap_opening_cost -150 \
    --gap_cost -50 --max_shift 1
```


### Python interface

The following code generates a bi-alignment of two toy proteins and shows
the resulting alignment in a graphical representation.

```python
import bialignment
import bialignment as ba
import timeit

args = {'type': 'Protein',
        'gap_cost': -50,
        'gap_opening_cost': -150,
        'shift_cost': -150,
        'structure_weight': 800,
        'max_shift': 1,
        'simmatrix': 'BLOSUM62'
       }

args['nameA'] = 'A'
args['nameB'] = 'B'
strA = "CHHHHHHHHHHHHHCCCCTCEEEEEEECCTCEEEEEEEECCC"
seqA = "RAKLPLKEKKLTATANYHPGIRYIMTGYSAKYIYSSTYARFR"
seqB = "KAKLPLKEKKLTRTANYHPGIRYIMTGYSAKRIYSSTYAYFR"
strB = "HHHHHHHHHHHHCCCCCCTCEEEEEEECCCCCEEEEEEEECC"


bialigner = ba.BiAligner(seqA, seqB, strA, strB,
                         **args)

score = bialigner.optimize()
print('SCORE',score)
print()

for line in bialigner.decode_trace():
    print(line)

ba.plot_alignment(bialigner.decode_trace_full(),
    width = 80,
    show_position_numbers=False,
    name_offset=3,
    #outname = "example.svg" #optionally write plot to file
    )
```
![](Examples/example.svg)


ChangeLog
0.3b (Nov 2021)
    * Implement efficient affine gap cost alignment
	  with O(n^2 max_shift^2) time complexity, 9 states and 15 cases per state
    * Preserve implementation of non-affine algorithm
	  (it is automatically chosen for gap_opening_cost 0)
    * Use sparse DP matrix O(n^2 max_shift^2)
    * Support similarity matrix for protein comparison, include BLOSUM62
    * Rewrite code for compilation with Cython and optimize due to static typing
    * Support input from files
    * Graphic representation of protein alignments
    * Smart trace back preferring more natural shifts (in case of
      ambiguity)
0.2 (Feb 14 2020)
    corresponding to the LNBI-CIBB post-proceedings submission
    * add version option
    * more flexible, prettier output of result bi-alignment
0.1 (May 03, 2019)
   corresponding to the CIBB conference submission
   *  stral-like scoring of structure
   *  highlighting in output
   *  basic bi-alignment functionality (treat bi-alignments like four-way
   *  alignments)
pre-versions:
    incomplete bi-alignment as three-way alignment


bialign_triplet.py
#!/usr/bin/env python3
import RNA

import sys
import argparse
import numpy as np
from math import log,exp

from bialign import *

## Alignment factory
class BiAlignerTriplet (BiAligner):
    def __init__(self, seqA, seqB, strA, strB, **params):
        super().__init__(seqA, seqB, strA, strB, **params)

    # iterator over recursion cases
    #
    # per case:
    #       pair of access info
    #     and
    #       function that returns list of case score components
    def recursionCases(self,i,j,k):
        mu1ij = self.mu1(i,j)
        mu2ik = self.mu2(i,k)
        Delta = self._params["shift_cost"]

        # synchronous cases
        yield ((1,1,1), mu1ij + mu2ik))
        yield ((1,0,0), self.g1A(i) + self.g2A(i))
        yield ((0,1,1), self.g1B(j) + self.g2B(k))
        # shifting
        yield ((1,1,0), mu1ij + self.g2A(i) + Delta)
        yield ((1,0,1), mu2ik + self.g1A(i) + Delta)
        yield ((0,1,0), self.g1B(j) + Delta)
        yield ((0,0,1), self.g2B(k) + Delta)

    def guardCase(self,x,i,j,k):
        (io,jo,ko) = x[0]
        return i-io>=0 and j-jo>=0 and k-ko>=0 and abs(k-ko-(j-jo))<=self._params["max_shift"]

    def evalCase(self,x,i,j,k):
        (io,jo,ko) = x[0]
        return self.M[i-io,j-jo,k-ko] + x[1]

    # run alignment algorithm
    def optimize(self):
        lenA = self.rnaA["len"]
        lenB = self.rnaB["len"]

        self.M = np.zeros((lenA+1,lenB+1,lenB+1), dtype=int)

        for i in range(0,lenA+1):
            for j in range(0,lenB+1):
                for k in range( max(0,j-self._params["max_shift"]), min(lenB+1, j+self._params["max_shift"]+1) ):
                    self.M[i,j,k] = self.plus( [ self.evalCase(x,i,j,k)
                                                 for x in self.recursionCases(i,j,k)
                                                 if self.guardCase(x,i,j,k) ] )
        return self.M[lenA,lenB,lenB]

    # perform traceback
    # @returns list of 'trace arrows'
    def traceback(self):
        lenA = self.rnaA["len"]
        lenB = self.rnaB["len"]

        trace=[]

        def trace_from(i,j,k):
            for x in self.recursionCases(i,j,k):
                if self.guardCase(x,i,j,k):
                    if self.evalCase(x,i,j,k) == self.M[i,j,k]:
                        (io,jo,ko) = x[0]
                        trace.append((io,jo,ko))
                        trace_from(i-io,j-jo,k-ko)
                        break

        trace_from(lenA,lenB,lenB)
        return list(reversed(trace))

    # decode trace to alignment strings
    def decode_trace(self,trace=None,show_structures=False):
        if trace is None:
            trace = self.traceback()

        rnas = (self.rnaA,self.rnaB,self.rnaB)
        pos = [0]*3
        alignment = [""]*3
        for i,y in enumerate(trace):
            for s in range(3):
                if (y[s]==0):
                    alignment[s] = alignment[s] + "-"
                elif (y[s]==1):
                    alignment[s] = alignment[s] + rnas[s]["seq"][pos[s]]
                    pos[s]+=1

        if not show_structures:
            return alignment

        # annotate with structure
        anno_alignment = list()
        for alistr,rna in zip(alignment,rnas):
            anno_alignment.append( self._transfer_gaps(alistr,rna["structure"]) )
            anno_alignment.append( alistr )

        return anno_alignment

    # evaluate trace
    def eval_trace(self,trace=None):
        if trace is None:
            trace = self.traceback()
        pos=[0]*3
        for i,y in enumerate(trace):
            for k in range(3): pos[k] += y[k]
            # lookup case
            for x in self.recursionCases(pos[0],pos[1],pos[2]):
                if x[0] == y:
                    line = " ".join([ str(x) for x in 
                            [pos,
                            y,
                            x[1],
                            "-->",
                            self.evalCase(x,pos[0],pos[1],pos[2])
                            ]
                        ])
                    yield line
                    break

#convenience function to construct bialigner, perform and return text output
def bialign_triplet(seqA,seqB,strA,strB,show_structures,verbose,**args):
    ba = BiAlignerTriplet(seqA,seqB,strA,strB,**args)

    optscore = ba.optimize()
    yield "SCORE:" + str(optscore)

    for s in ba.decode_trace(show_structures=show_structures):
        yield s

    if verbose:
        yield from ba.eval_trace()

def main():
    parser = argparse.ArgumentParser(description= "Bialignment.")
    add_bialign_parameters(parser)
    
    args = parser.parse_args()
    print("Args",vars(args))

    for line in bialign_triplet(**vars(args)):
        print(line)

if __name__ == "__main__":
    main()


bialignment_nonpyx.py
from collections import defaultdict

blosum62 = """-  A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X  *
A  4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0 -2 -1  0 -4
R -1  5  0 -2 -3  1  0 -2  0 -3 -2  2 -1 -3 -2 -1 -1 -3 -2 -3 -1  0 -1 -4
N -2  0  6  1 -3  0  0  0  1 -3 -3  0 -2 -3 -2  1  0 -4 -2 -3  3  0 -1 -4
D -2 -2  1  6 -3  0  2 -1 -1 -3 -4 -1 -3 -3 -1  0 -1 -4 -3 -3  4  1 -1 -4
C  0 -3 -3 -3  9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1 -3 -3 -2 -4
Q -1  1  0  0 -3  5  2 -2  0 -3 -2  1  0 -3 -1  0 -1 -2 -1 -2  0  3 -1 -4
E -1  0  0  2 -4  2  5 -2  0 -3 -3  1 -2 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4
G  0 -2  0 -1 -3 -2 -2  6 -2 -4 -4 -2 -3 -3 -2  0 -2 -2 -3 -3 -1 -2 -1 -4
H -2  0  1 -1 -3  0  0 -2  8 -3 -3 -1 -2 -1 -2 -1 -2 -2  2 -3  0  0 -1 -4
I -1 -3 -3 -3 -1 -3 -3 -4 -3  4  2 -3  1  0 -3 -2 -1 -3 -1  3 -3 -3 -1 -4
L -1 -2 -3 -4 -1 -2 -3 -4 -3  2  4 -2  2  0 -3 -2 -1 -2 -1  1 -4 -3 -1 -4
K -1  2  0 -1 -3  1  1 -2 -1 -3 -2  5 -1 -3 -1  0 -1 -3 -2 -2  0  1 -1 -4
M -1 -1 -2 -3 -1  0 -2 -3 -2  1  2 -1  5  0 -2 -1 -1 -1 -1  1 -3 -1 -1 -4
F -2 -3 -3 -3 -2 -3 -3 -3 -1  0  0 -3  0  6 -4 -2 -2  1  3 -1 -3 -3 -1 -4
P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3 -3 -1 -2 -4  7 -1 -1 -4 -3 -2 -2 -1 -2 -4
S  1 -1  1  0 -1  0  0  0 -1 -2 -2  0 -1 -2 -1  4  1 -3 -2 -2  0  0  0 -4
T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1  1  5 -2 -2  0 -1 -1  0 -4
W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3 -2 -3 -1  1 -4 -3 -2 11  2 -3 -4 -3 -2 -4
Y -2 -2 -2 -3 -2 -1 -2 -3  2 -1 -1 -2 -1  3 -3 -2 -2  2  7 -1 -3 -2 -1 -4
V  0 -3 -3 -3 -1 -2 -2 -3 -3  3  1 -2  1 -1 -2 -2  0 -3 -1  4 -3 -2 -1 -4
B -2 -1  3  4 -3  0  1 -1  0 -3 -4  0 -3 -3 -2  0 -1 -4 -3 -3  4  1 -1 -4
Z -1  0  0  1 -3  3  4 -2  0 -3 -3  1 -1 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4
X  0 -1 -1 -1 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -2  0  0 -2 -1 -1 -1 -1 -1 -4
* -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4  1
"""


def read_simmatrix(filename, scale=100):
    if filename == "BLOSUM62":
        lines = blosum62.split("\n")
    else:
        with open(filename, "r") as fh:
            lines = fh.readlines()

    keys = None
    keys2 = []
    matrix = dict()

    for i, line in enumerate(lines):
        if keys and i > len(keys):
            break
        line = line.split()
        if line[0] == "-":
            keys = line[1:]
        else:
            keys2.append(line[0])
            matrix[line[0]] = {
                key: (scale * int(val)) for key, val in zip(keys, line[1:])
            }

    if not keys == keys2:
        print("ERROR while reading simmatrix {filename}.")
    return matrix


def read_molecule(content, type):
    if type != "Protein":
        raise IOError(f"Cannot read files of type {type}")

    result = defaultdict(lambda: "")
    keys = ["Query", "Struc"]
    for line in content.split("\n"):
        line = line.split()
        if not line:
            continue
        if line[0] in keys:
            if len(line) != 4:
                raise IOError("Cannot parse")
            result[line[0]] += line[2]

    if len(result[keys[0]]) != len(result[keys[1]]):
        raise IOError("Sequence and structure of unequal length.")
    if len(result[keys[0]]) == 0:
        raise IOError("Input does not contain input sequence and structure.")

    return [result[k] for k in keys]


def read_molecule_from_file(filename, type):
    try:
        with open(filename, "r") as fh:
            return read_molecule(fh.read(), type)
    except FileNotFoundError as e:
        print("Input file not found.")
        print(e)
        sys.exit(-1)
    except IOError as e:
        print(f"Cannot read input file {filename}.")
        print(e)
        sys.exit(-1)


def breaklines(alilines, width):
    res = []

    offset = 0
    length = len(alilines[0][1])

    while offset < length:
        resblock = []
        for i, line in enumerate(alilines):
            name = line[0]
            aliline = line[1][offset : offset + width]
            resblock.append((name, aliline))
        offset += width

        res.append(resblock)

    return res


def runs(s):
    if s == "":
        return

    last_start = 0
    last = s[0]
    for i, x in enumerate(s[1:]):
        if x != last:
            yield (last, last_start, i + 1)
            last_start = i + 1
            last = x
    yield (last, last_start, len(s))


# struct = "HHHCCC---TTTCCHHHHHTT--EEECC"
# print(len(struct))
# for c,s,e in runs(struct):
#    print(c,s,e,struct[s:e])

helix_yadd_a = []
helix_yadd_b = []


def fourway_from_full(alilines):
    return [alilines[i] for i in [1, 3, 6, 8, 12, 13]]


def plot_alignment(
    alilines,
    width,
    *,
    show_structure_strings=False,
    name_offset=12,
    show_position_numbers=True,
    outname=None,
):

    import matplotlib.pyplot as plt
    from matplotlib.patches import Rectangle
    from collections import defaultdict

    helix_yadd_a = [0.0075]
    helix_yadd_b = [0.0075]

    if len(alilines) >= 13:  # full alilines
        alilines = fourway_from_full(alilines)

    aliblocks = breaklines(alilines, width)

    numblocks = len(aliblocks)
    fig, axs = plt.subplots(numblocks, 1, figsize=(0.18 * width, 2 * numblocks))

    if numblocks == 1:
        axs = [axs]

    font = {"family": "monospace", "weight": "normal", "size": 16.0}
    plt.rc("font", **font)  # pass in the font dict as kwargs

    colors = defaultdict(lambda: "grey", E="green", C="orange", T="blue", H="red")
    colors["-"] = None

    def draw_line(ax, s, e, y, color, lw):
        ax.plot(
            [s, e],
            [y + 0.025, y + 0.025],
            linewidth=lw,
            color=color,
            solid_capstyle="butt",
        )
        # lwcorr = (8-lw)/20
        # ax.plot([s+0.48-lwcorr,e-0.36+lwcorr], [y+0.025,y+0.025], linewidth=lw, color=color)

    def draw_sheet(ax, s, e, y, color):
        if s + 1 < e:
            ax.plot(
                [s, e - 1],
                [y + 0.025, y + 0.025],
                linewidth=8,
                color=color,
                solid_capstyle="butt",
            )
        ax.plot(
            [e - 0.05], [y + 0.025], linewidth=0, color=color, marker=5, markersize=13
        )

    def draw_helix(ax, s, e, y, color, yadd):
        xs = list(reversed(range(s, e + 1)))
        y += 0.025
        ys = [y + yadd[0]]
        for i in reversed(range(s, e)):
            yadd[0] = -yadd[0]
            ys.append(y + yadd[0])

        ax.plot(
            xs,
            ys,
            linewidth=6,
            color=color,
            solid_capstyle="butt",
            solid_joinstyle="round",
        )

    def draw_str(ax, y, line, yadd):
        name, struc = line
        for ch, s, e in reversed(list(runs(struc))):
            color = colors[ch]
            if ch == "E":
                draw_sheet(ax, s, e, y, color)
            elif ch == "H":
                draw_helix(ax, s, e, y, color, yadd)
            elif color == None:
                pass
            else:
                if ch == "T":
                    w = 8
                else:
                    w = 4
                draw_line(ax, s, e, y, color, w)

    def draw_seq(ax, y, line, other=None):
        name, seq = line
        ax.text(-name_offset, y, name)
        for x, ch in enumerate(seq):
            weight = "normal"
            color = "black"
            if other is not None:
                if ch != "-" and other[x] != "-":
                    color = "darkred"
                if ch == other[x]:
                    weight = "bold"
                    color = "black"
            ax.text(x, y, ch, weight=weight, color=color)

    def draw_shifts(ax, aa, bb):
        for x, (a, b) in enumerate(zip(aa, bb)):
            if a in ["<", ">"] or b in ["<", ">"]:
                ax.add_patch(
                    Rectangle(
                        (x, -0.022), 1, 0.4, edgecolor="black", fill=False, lw=0.5
                    )
                )

    offset_a = 1
    offset_b = 1

    for k, block in enumerate(aliblocks):
        ax = axs[k]
        ax.set_xlim(-0.5, width + 0.5)
        ax.set_ylim(-0.175, 0.425)
        ax.axis("off")

        # print(k, block)

        # for i, (name, aliline) in enumerate(block):
        #    print(f"{i:2} {name:12} {aliline}")
        # print()

        length = len(block[0][1])
        length_a = len(block[0][1].replace("-", ""))
        length_b = len(block[1][1].replace("-", ""))

        ## print sequence positions
        if show_position_numbers:
            ax.text(0, 0.435, offset_a, fontsize=10)
            offset_a += length_a
            ax.text(length, 0.435, offset_a - 1, fontsize=10, ha="right")

            ax.text(0, -0.12, offset_b, fontsize=10)
            offset_b += length_b
            ax.text(length, -0.12, offset_b - 1, fontsize=10, ha="right")

        draw_seq(ax, 0.2, block[0], block[1][1])
        draw_seq(ax, 0.1, block[1], block[0][1])

        draw_str(ax, 0.3, block[2], helix_yadd_a)
        draw_str(ax, 0.025, block[3], helix_yadd_b)
        if show_structure_strings:
            draw_seq(ax, 0.3, ("", block[2][1]))
            draw_seq(ax, 0, ("", block[3][1]))

    if len(block) > 4:
        for k, block in enumerate(aliblocks):
            ax = axs[k]
            draw_seq(ax, 0.375, ("", block[4][1].replace(".", " ")))
            draw_seq(ax, -0.075, ("", block[5][1].replace(".", " ")))
            draw_shifts(ax, block[4][1], block[5][1])

    if outname is not None:
        plt.savefig(outname)
    plt.show()


bialign.py
#!/usr/bin/env python3
import argparse

import bialignment
import sys

VERSION_STRING = f"BiAlign {bialignment.__version__}"


def bialign(seqA, seqB, strA, strB, verbose, **args):
    ba = bialignment.BiAligner(seqA, seqB, strA, strB, **args)

    optscore = ba.optimize()
    yield "SCORE: " + str(optscore)
    yield ""

    ali = ba.decode_trace()

    yield from ali

    if verbose:
        yield from ba.eval_trace()


def add_bialign_parameters(parser):
    parser.add_argument("seqA", help="sequence A")
    parser.add_argument("seqB", help="sequence B")
    parser.add_argument("--strA", default=None, help="structure A")
    parser.add_argument("--strB", default=None, help="structure B")
    parser.add_argument("--nameA", default="A", help="name A")
    parser.add_argument("--nameB", default="B", help="name B")
    parser.add_argument("-v", "--verbose", action="store_true", help="Verbose")

    parser.add_argument(
        "--type", default="RNA", type=str, help="Type of molecule: RNA or Protein"
    )

    parser.add_argument(
        "--nodescription",
        action="store_true",
        help="Don't prefix the strings in output alignment with descriptions",
    )
    parser.add_argument(
        "--outmode",
        default="default",
        help="Output mode [call --mode help for a list of options]",
    )

    parser.add_argument(
        "--sequence_match_similarity",
        type=int,
        default=100,
        help="Similarity of matching nucleotides",
    )
    parser.add_argument(
        "--sequence_mismatch_similarity",
        type=int,
        default=0,
        help="Similarity of mismatching nucleotides",
    )
    parser.add_argument(
        "--structure_weight",
        type=int,
        default=400,
        help="Weighting factor for structure similarity",
    )
    parser.add_argument(
        "--gap_opening_cost",
        type=int,
        default=0,
        help="Similarity of opening a gap (turns on affine gap cost if not 0)",
    )
    parser.add_argument(
        "--gap_cost", type=int, default=-200, help="Similarity of a single gap position"
    )
    parser.add_argument(
        "--shift_cost",
        type=int,
        default=-250,
        help="Similarity of shifting the two scores against each other",
    )
    parser.add_argument(
        "--max_shift",
        type=int,
        default=2,
        help="Maximal number of shifts away from the diagonal in either direction",
    )
    parser.add_argument(
        "--fileinput",
        action="store_true",
        help="Read sequence and structure input from file",
    )

    parser.add_argument("--version", action="version", version=VERSION_STRING)

    parser.add_argument("--simmatrix", type=str, default=None, help="Similarity matrix")


def main():
    parser = argparse.ArgumentParser(description="Bialignment.")
    add_bialign_parameters(parser)

    args = parser.parse_args()

    if args.fileinput:
        args.seqA, args.strA = bialignment.read_molecule_from_file(args.seqA, args.type)
        args.seqB, args.strB = bialignment.read_molecule_from_file(args.seqB, args.type)

    input_descr = ["Input:", "seqA\t " + args.seqA, "seqB\t " + args.seqB]
    if hasattr(args, "strA") and args.strA is not None:
        input_descr.append("strA\t " + args.strA)
    if hasattr(args, "strB") and args.strB is not None:
        input_descr.append("strB\t " + args.strB)

    print("\n".join(input_descr))

    if args.outmode == "help":
        print()
        print("Available modes: " + ", ".join(bialignment.BiAligner.outmodes.keys()))
        print()
        exit()

    for line in bialign(**vars(args)):
        print(line)


if __name__ == "__main__":
    main()
